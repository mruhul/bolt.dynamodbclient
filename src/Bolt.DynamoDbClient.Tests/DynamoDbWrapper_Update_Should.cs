using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using Bolt.DynamoDbClient.Tests.Builders;
using Bolt.DynamoDbClient.Tests.TestHelpers;

namespace Bolt.DynamoDbClient.Tests
{
    public class DynamoDbWrapper_Update_Should
    {
        private IDynamoDbWrapper sut;
        private IAmazonDynamoDB fake;

        public DynamoDbWrapper_Update_Should()
        {
            fake = Substitute.For<IAmazonDynamoDB>();
            sut = DynamoDbWrapperBuilder.Build(fake);
        }

        [Theory]
        [MemberData(nameof(TestData))]
        public void call_dynamo_with_correct_request_when_doesnt_skip_null_value(TestData<ComplexRecord> input)
        {
            UpdateItemRequest gotRequest = null;

            fake.UpdateItemAsync(Arg.Any<UpdateItemRequest>(), Arg.Any<CancellationToken>())
                .Returns(new UpdateItemResponse())
                .AndDoes(c =>
                {
                    gotRequest = c.Arg<UpdateItemRequest>();
                });


            sut.Update(input.Value, false);

            fake.Received().UpdateItemAsync(Arg.Any<UpdateItemRequest>(), Arg.Any<CancellationToken>());

            new
            {
                Key = input.Key,
                Scenario = input.Scenario,
                GotRequest = gotRequest
            }.ShouldMatchApproved(input.Key, input.Scenario);
        }

        [Theory]
        [MemberData(nameof(TestData))]
        public void call_dynamo_with_correct_request_when_skip_null_value(TestData<ComplexRecord> input)
        {
            UpdateItemRequest? gotRequest = null;

            fake.UpdateItemAsync(Arg.Any<UpdateItemRequest>(), Arg.Any<CancellationToken>())
                .Returns(new UpdateItemResponse())
                .AndDoes(c =>
                {
                    gotRequest = c.Arg<UpdateItemRequest>();
                });


            sut.Update(input.Value, true);

            fake.Received().UpdateItemAsync(Arg.Any<UpdateItemRequest>(), Arg.Any<CancellationToken>());

            new
            {
                Key = input.Key,
                Scenario = input.Scenario,
                GotRequest = gotRequest
            }.ShouldMatchApproved(input.Key, input.Scenario);
        }

        public static IEnumerable<object[]> TestData = new TestData<ComplexRecord>[]
        {
            new()
            {
                Key = "1",
                Scenario = "When all value provided",
                Value = ComplexRecordBuilder.Build()
            },
            new()
            {
                Key = "2",
                Scenario = "When nullable are empty",
                Value = ComplexRecordBuilder.Build() with
                {
                    DecimalNullableValue = null,
                    FloatNullableValue = null,
                    GuildNullableValue = null,
                    IntNullableValue= null,
                    NullableStringValue = null,
                }
            },
        }.ToTestData();
    }
}
