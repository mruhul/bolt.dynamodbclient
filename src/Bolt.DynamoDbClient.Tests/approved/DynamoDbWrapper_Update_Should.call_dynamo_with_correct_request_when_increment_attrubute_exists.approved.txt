{
  "attributeUpdates": {},
  "expected": {},
  "expressionAttributeNames": {
    "#Id": "Id",
    "#Name": "Name",
    "#LikeCount": "LikeCount",
    "#PhotoCount": "PhotoCount",
    "#VideoCount": "VideoCount"
  },
  "expressionAttributeValues": {
    ":Id": {
      "bool": false,
      "isBOOLSet": false,
      "bs": [],
      "isBSSet": false,
      "l": [],
      "isLSet": false,
      "m": {},
      "isMSet": false,
      "ns": [],
      "isNSSet": false,
      "null": false,
      "s": "3fb8b9df-bab9-4031-b2ae-99b8f04a4940",
      "ss": [],
      "isSSSet": false
    },
    ":Name": {
      "bool": false,
      "isBOOLSet": false,
      "bs": [],
      "isBSSet": false,
      "l": [],
      "isLSet": false,
      "m": {},
      "isMSet": false,
      "ns": [],
      "isNSSet": false,
      "null": false,
      "s": "first last",
      "ss": [],
      "isSSSet": false
    },
    ":LikeCount": {
      "bool": false,
      "isBOOLSet": false,
      "bs": [],
      "isBSSet": false,
      "l": [],
      "isLSet": false,
      "m": {},
      "isMSet": false,
      "n": "1",
      "ns": [],
      "isNSSet": false,
      "null": false,
      "ss": [],
      "isSSSet": false
    },
    ":start": {
      "bool": false,
      "isBOOLSet": false,
      "bs": [],
      "isBSSet": false,
      "l": [],
      "isLSet": false,
      "m": {},
      "isMSet": false,
      "n": "0",
      "ns": [],
      "isNSSet": false,
      "null": false,
      "ss": [],
      "isSSSet": false
    },
    ":PhotoCount": {
      "bool": false,
      "isBOOLSet": false,
      "bs": [],
      "isBSSet": false,
      "l": [],
      "isLSet": false,
      "m": {},
      "isMSet": false,
      "n": "2",
      "ns": [],
      "isNSSet": false,
      "null": false,
      "ss": [],
      "isSSSet": false
    },
    ":VideoCount": {
      "bool": false,
      "isBOOLSet": false,
      "bs": [],
      "isBSSet": false,
      "l": [],
      "isLSet": false,
      "m": {},
      "isMSet": false,
      "n": "-1",
      "ns": [],
      "isNSSet": false,
      "null": false,
      "ss": [],
      "isSSSet": false
    }
  },
  "key": {
    "PK": {
      "bool": false,
      "isBOOLSet": false,
      "bs": [],
      "isBSSet": false,
      "l": [],
      "isLSet": false,
      "m": {},
      "isMSet": false,
      "ns": [],
      "isNSSet": false,
      "null": false,
      "s": "my-pk",
      "ss": [],
      "isSSSet": false
    },
    "SK": {
      "bool": false,
      "isBOOLSet": false,
      "bs": [],
      "isBSSet": false,
      "l": [],
      "isLSet": false,
      "m": {},
      "isMSet": false,
      "ns": [],
      "isNSSet": false,
      "null": false,
      "s": "my-sk",
      "ss": [],
      "isSSSet": false
    }
  },
  "tableName": "SummaryRecord",
  "updateExpression": "SET #Id = :Id, #Name = :Name, #LikeCount = if_not_exists(#LikeCount, :start) \u002B :LikeCount, #PhotoCount = if_not_exists(#PhotoCount, :start) \u002B :PhotoCount, #VideoCount = if_not_exists(#VideoCount, :start) \u002B :VideoCount"
}